File: flag.py
import math
import random
import sys
import re

NONE = 0
CMP  = 1 # type supports equality and inequelity operators
TCMP = 3 # type supports all comparison operators
HASH = 5 # type can be hashed
DEFC = 8 # default constructor
ALL  = TCMP | HASH | DEFC

# definisci la curva di probabilita'
# f(x) = e^kx
# integrale di f(x) = e^kx / k
# e^kx / k = y
# inverso dell'integrale
# x = log(y * k)/k

def choose(k, s):
    if k == 0:
        return random.choice(s)

    # integrale definito in [0, n]
    a, b = math.exp(k * 0) / k, math.exp(k * len(s)) / k
    y = random.uniform(a, b)
    # funzione inversa dell'integrale
    x = math.log(y * k) / k
    return s[int(x)]

def cs(flags2, flags3):
    def f(k, flags4):
        flags = flags2 | flags3 & flags4
        l = [(g, f) for op, g, f in TEMPLATES if op & flags == flags]
        gens, f = choose(k, l)
        return f.format(*(g(k - 0.10 - random.uniform(0.01, 0.1), flags) for g in gens))
    return f

def craft_flag(n : int = 10):
    return f'`{cs(NONE, ALL)(0.2, DEFC)} flag[{n}];`'

TEMPLATES = [
    (ALL , [], 'int'),
    (ALL , [], 'size_t'),
    (ALL , [], 'uint32_t'),
    (ALL , [], 'int32_t'),
    (ALL , [], 'll'),
    (ALL , [], 'unsigned ll int'),
    (TCMP | DEFC , [], '__int128_t'),
    (ALL , [], 'uint64_t'),
    (ALL , [], 'int64_t'),
    (DEFC, [], 'std::mt19937_64'),
    (ALL , [], 'std::string'),
    (DEFC, [], 'std::any'),
    # (NONE, [], 'null_type'),
    (ALL , [], 'bool'),
    (ALL , [], 'std::byte'),
    (ALL , [], 'signed char'),
    (ALL , [], 'unsigned char'),
    (CMP , [], 'std::strong_ordering'),
    (CMP , [], 'std::weak_ordering'),
    (CMP , [], 'std::partial_ordering'),
    (DEFC, [], 'std::mutex'),
    (DEFC, [], 'std::thread'),
    (CMP | HASH | DEFC, [], 'std::bitset<262144>'),
    (TCMP | DEFC, [], 'std::chrono::seconds'),
    (DEFC, [], 'FILE'),
    (DEFC, [], 'std::ifstream'),
    (DEFC, [], 'std::ofstream'),
    (ALL , [cs(NONE, NONE)], '{} *'),
    # (ALL , [cs(NONE, ALL)], '{}&'),
    # (ALL , [cs(NONE, ALL)], '{}[]'),
    # (ALL , [cs(NONE, ALL)], '{} const'),
    (TCMP | DEFC, [cs(NONE, TCMP)], 'std::deque<{}>'),
    (TCMP | DEFC, [cs(TCMP, NONE)], 'std::set<{}>'),
    (TCMP | DEFC, [cs(TCMP, NONE)], 'std::multiset<{}>'),
    (CMP | DEFC, [cs(HASH, NONE)], 'std::unordered_set<{}>'),
    (CMP | DEFC, [cs(HASH, NONE)], 'std::unordered_multiset<{}>'),
    (DEFC, [cs(TCMP, NONE)], 'std::priority_queue<{}>'),
    (TCMP | DEFC, [cs(NONE, TCMP | DEFC)], 'std::array<{}, 2>'),
    (TCMP | DEFC, [cs(NONE, TCMP | DEFC)], 'std::array<{}, 3>'),
    (TCMP | DEFC, [cs(NONE, TCMP)], 'std::vector<{}>'),
    (TCMP | DEFC, [cs(NONE, NONE)], 'std::unique_ptr<{}>'),
    (ALL , [cs(NONE, NONE)], 'std::shared_ptr<{}>'),
    (DEFC, [cs(NONE, NONE)], 'std::weak_ptr<{}>'),
    # (NONE, [cs(NONE, ALL)], 'std::auto_ptr<{}>'),
    (ALL , [cs(NONE, TCMP | HASH)], 'std::optional<{}>'),
    (DEFC, [cs(NONE, ALL)], 'std::span<{}>'),
    (DEFC, [cs(TCMP, NONE)], 'std::priority_queue<{0}, std::vector<{0}>, std::greater<{0}>>'),
    # (NONE, [cs(NONE, ALL)], 'std::chrono::time_point<std::chrono::high_resolution_clock, std::chrono::duration<{}, std::ratio<1, 1000000000>>>'),
    (DEFC, [cs(TCMP, NONE)], '__gnu_pbds::tree<{0}, __gnu_pbds::null_type, std::less<{0}>, __gnu_pbds::rb_tree_tag, __gnu_pbds::tree_order_statistics_node_update>'),
    (TCMP | DEFC, [cs(NONE, ALL), cs(NONE, ALL)], 'std::pair<{}, {}>'),
    (ALL , [cs(NONE, TCMP | HASH), cs(NONE, TCMP | HASH)], 'std::variant<{}, {}>'),
    (TCMP | DEFC, [cs(TCMP, NONE), cs(NONE, TCMP)], 'std::map<{}, {}>'),
    (TCMP | DEFC, [cs(TCMP, NONE), cs(NONE, TCMP)], 'std::multimap<{}, {}>'),
    (CMP | DEFC, [cs(HASH, NONE), cs(NONE, CMP)], 'std::unordered_map<{}, {}>'),
    (CMP | DEFC, [cs(HASH, NONE), cs(NONE, CMP)], 'std::unordered_multimap<{}, {}>'),
    (ALL , [cs(NONE, ALL), cs(NONE, NONE)], 'std::conditional_t<true, {}, {}>'),
    (ALL , [cs(NONE, NONE), cs(NONE, ALL)], 'std::conditional_t<false, {}, {}>'),
    (TCMP | DEFC, [cs(NONE, ALL), cs(NONE, ALL), cs(NONE, ALL)], 'std::tuple<{}, {}, {}>'),
    (TCMP | DEFC, [cs(NONE, ALL), cs(NONE, ALL), cs(NONE, ALL), cs(NONE, ALL)], 'std::tuple<{}, {}, {}, {}>'),
]

if __name__ == '__main__':
    if (len(sys.argv) > 1):
        if re.match(r'^\[\d+\]$', sys.argv[1]):
            n = int(re.search(r'^\[(\d+)\]$', sys.argv[1]).group(1))
            print(craft_flag(n))
        else:
            print('Usage: python3 flag.py [N]')
            print('N is the number of flags to generate')
            print('If N is not specified, 10 flags will be generated')
            exit(0)
    else:
        print(craft_flag())

File: src/flag.rs
const NONE: u32 = 0;
const CMP: u32 = 1; // type supports equality and inequelity operators
const TCMP: u32 = 3; // type supports all comparison operators
const HASH: u32 = 5; // type can be hashed
const DEFC: u32 = 8; // default constructor
const ALL: u32 = TCMP | HASH | DEFC;

#[derive(Debug, Clone, Copy)]
enum X {
    Str(&'static str),
    Gen(u32, u32),
}

fn choose<T>(k: f64, mut it: impl Iterator<Item = T>) -> Option<T> {
    let mut val = it.next()?;
    let k = k.exp();
    let mut p = 1.0;
    let mut total = 0.0;
    for val2 in it {
        total += p;
        p *= k;
        if rand::random::<f64>() * total < p {
            val = val2;
        }
    }
    Some(val)
}

fn gen_type(out: &mut String, k: f64, flags: u32) {
    let it = TEMPLATES.iter().filter(|(f, _)| f & flags == flags);
    let &(_, gens) = choose(k, it).expect("No template found");
    for &gen in gens {
        match gen {
            X::Str(s) => out.push_str(s),
            X::Gen(or, and) => {
                let del = rand::random::<f64>() * (0.20 - 0.10) + 0.10;
                gen_type(out, k - del, or | and & flags);
            }
        }
    }
}

pub async fn get_flag() -> String {
    let mut out = String::new();
    out.push('`');
    gen_type(&mut out, 0.2, DEFC);
    out.push_str(" flag[10];`");
    out
}

static TEMPLATES: &[(u32, &[X])] = &[
    (ALL, &[X::Str("int")]),
    (ALL, &[X::Str("size_t")]),
    (ALL, &[X::Str("uint32_t")]),
    (ALL, &[X::Str("int32_t")]),
    (ALL, &[X::Str("ll")]),
    (ALL, &[X::Str("unsigned ll int")]),
    (TCMP | DEFC, &[X::Str("__int128_t")]),
    (ALL, &[X::Str("uint64_t")]),
    (ALL, &[X::Str("int64_t")]),
    (DEFC, &[X::Str("std::mt19937_64")]),
    (ALL, &[X::Str("std::string")]),
    (DEFC, &[X::Str("std::any")]),
    (ALL, &[X::Str("bool")]),
    (ALL, &[X::Str("std::byte")]),
    (ALL, &[X::Str("signed char")]),
    (ALL, &[X::Str("unsigned char")]),
    (CMP, &[X::Str("std::strong_ordering")]),
    (CMP, &[X::Str("std::weak_ordering")]),
    (CMP, &[X::Str("std::partial_ordering")]),
    (DEFC, &[X::Str("std::mutex")]),
    (DEFC, &[X::Str("std::thread")]),
    (CMP | HASH | DEFC, &[X::Str("std::bitset<262144>")]),
    (TCMP | DEFC, &[X::Str("std::chrono::seconds")]),
    (DEFC, &[X::Str("FILE")]),
    (DEFC, &[X::Str("std::ifstream")]),
    (DEFC, &[X::Str("std::ofstream")]),
    (ALL, &[X::Gen(NONE, NONE), X::Str(" *")]),
    (
        TCMP | DEFC,
        &[X::Str("std::deque<"), X::Gen(NONE, TCMP), X::Str(">")],
    ),
    (
        TCMP | DEFC,
        &[X::Str("std::set<"), X::Gen(TCMP, NONE), X::Str(">")],
    ),
    (
        TCMP | DEFC,
        &[X::Str("std::multiset<"), X::Gen(TCMP, NONE), X::Str(">")],
    ),
    (
        CMP | DEFC,
        &[
            X::Str("std::unordered_set<"),
            X::Gen(HASH, NONE),
            X::Str(">"),
        ],
    ),
    (
        CMP | DEFC,
        &[
            X::Str("std::unordered_multiset<"),
            X::Gen(HASH, NONE),
            X::Str(">"),
        ],
    ),
    (
        DEFC,
        &[
            X::Str("std::priority_queue<"),
            X::Gen(TCMP, NONE),
            X::Str(">"),
        ],
    ),
    (
        TCMP | DEFC,
        &[
            X::Str("std::array<"),
            X::Gen(NONE, TCMP | DEFC),
            X::Str(", 2>"),
        ],
    ),
    (
        TCMP | DEFC,
        &[
            X::Str("std::array<"),
            X::Gen(NONE, TCMP | DEFC),
            X::Str(", 3>"),
        ],
    ),
    (
        TCMP | DEFC,
        &[X::Str("std::vector<"), X::Gen(NONE, TCMP), X::Str(">")],
    ),
    (
        TCMP | DEFC,
        &[X::Str("std::unique_ptr<"), X::Gen(NONE, NONE), X::Str(">")],
    ),
    (
        ALL,
        &[X::Str("std::shared_ptr<"), X::Gen(NONE, NONE), X::Str(">")],
    ),
    (
        DEFC,
        &[X::Str("std::weak_ptr<"), X::Gen(NONE, NONE), X::Str(">")],
    ),
    (
        ALL,
        &[
            X::Str("std::optional<"),
            X::Gen(NONE, TCMP | HASH),
            X::Str(">"),
        ],
    ),
    (
        DEFC,
        &[
            X::Str("std::span<"),
            X::Gen(NONE, ALL),
            X::Str(", std::dynamic_extent>"),
        ],
    ),
    (
        TCMP | DEFC,
        &[
            X::Str("std::pair<"),
            X::Gen(NONE, ALL),
            X::Str(", "),
            X::Gen(NONE, ALL),
            X::Str(">"),
        ],
    ),
    (
        ALL,
        &[
            X::Str("std::variant<"),
            X::Gen(NONE, TCMP | HASH),
            X::Str(", "),
            X::Gen(NONE, TCMP | HASH),
            X::Str(">"),
        ],
    ),
    (
        ALL,
        &[
            X::Str("std::variant<"),
            X::Gen(NONE, TCMP | HASH),
            X::Str(", "),
            X::Gen(NONE, TCMP | HASH),
            X::Str(", "),
            X::Gen(NONE, TCMP | HASH),
            X::Str(">"),
        ],
    ),
    (
        TCMP | DEFC,
        &[
            X::Str("std::map<"),
            X::Gen(TCMP, NONE),
            X::Str(", "),
            X::Gen(NONE, TCMP),
            X::Str(">"),
        ],
    ),
    (
        TCMP | DEFC,
        &[
            X::Str("std::multimap<"),
            X::Gen(TCMP, NONE),
            X::Str(", "),
            X::Gen(NONE, TCMP),
            X::Str(">"),
        ],
    ),
    (
        CMP | DEFC,
        &[
            X::Str("std::unordered_map<"),
            X::Gen(HASH, NONE),
            X::Str(", "),
            X::Gen(NONE, CMP),
            X::Str(">"),
        ],
    ),
    (
        CMP | DEFC,
        &[
            X::Str("std::unordered_multimap<"),
            X::Gen(HASH, NONE),
            X::Str(", "),
            X::Gen(NONE, CMP),
            X::Str(">"),
        ],
    ),
    (
        ALL,
        &[
            X::Str("std::conditional_t<true, "),
            X::Gen(NONE, ALL),
            X::Str(", "),
            X::Gen(NONE, ALL),
            X::Str(">"),
        ],
    ),
    (
        ALL,
        &[
            X::Str("std::conditional_t<false, "),
            X::Gen(NONE, ALL),
            X::Str(", "),
            X::Gen(NONE, ALL),
            X::Str(">"),
        ],
    ),
    (
        TCMP | DEFC,
        &[
            X::Str("std::tuple<"),
            X::Gen(NONE, ALL),
            X::Str(", "),
            X::Gen(NONE, ALL),
            X::Str(", "),
            X::Gen(NONE, ALL),
            X::Str(">"),
        ],
    ),
    (
        TCMP | DEFC,
        &[
            X::Str("std::tuple<"),
            X::Gen(NONE, ALL),
            X::Str(", "),
            X::Gen(NONE, ALL),
            X::Str(", "),
            X::Gen(NONE, ALL),
            X::Str(", "),
            X::Gen(NONE, ALL),
            X::Str(">"),
        ],
    ),
];
